docs/common-sdk-design.md

```diff
@@ -0,0 +1,351 @@
+# Common SDK infrastructure dev design
+The purpose of this doc is to serve as a guidance for software engineers to implement a solution defined below as "the common SDK infrastructure". It contains basic analysis of the problem, proposal for improvements, goals, requirements, architecture and a list of technical challenges with suggestions for implementation.
+
+## Problem analysis
+There is nothing fundamentally wrong with current PlayFab SDKs and the SDK generation process. They are technically designed and implemented well, relatively simple, efficient and address most of our today's customers' needs. The simplicity of current implementation - while being a powerful benefit - also has a cost. There are certain limitations that may be standing in a way of even greater adoption of the platform by customers with specific needs or extending business opportunities provided by the platform:
+* Monolithic, one-for-all formation of current SDKs results in:
+    * Limited configurability (ability to select which components are needed or should be excluded from the SDK)
+    * Limited customization options (ability to easily replace standard components with custom ones in the SDK)
+    * Limited extensibility (ability to easily add new external functionality to the SDK)
+* No possibility to build a custom SDK
+* No easy and configurable way to:
+    * Use an alternative service platform (e.g. a different data ingestion system)
+    * Use a different network transport/protocol (unless already built in the SDK)
+    * Use different elements of data processing stack on the client side (e.g. specific data formatting, serialization, compression, encryption)
+    * Use alternative client-side API
+*  No diversity in available solutions (beside a single SDK (per client platform) and REST API) that would allow a wider spectrum of integration options with PlayFab services (think of a "marketplace" of various solutions created by PlayFab and third parties, that would allow users to choose the breadth and the depth of interaction with PlayFab platform)
+
+## Solution
+Many limitations listed above can be addressed by a set of improvements in SDKs themselves and in the process of their generation:
+
+* SDKs:
+    * Modular composition (SDK consists of smaller building blocks - _plugins_)
+    * Dynamic formation (plugins can be removed, added, replaced) based on SDK configuration
+    * Custom plugins (created by users) are allowed
+    * Plugins can interact with other plugins
```

Comment:
>Plugins can interact with other plugins [](start = 6, length = 39)

Does this mean data is never brokered by the Plugin manager or is it sometimes ?

Reply:
At this evolutionary step it doesn't. The goal and purpose of Plugin Manager is to manage plugins - i.e. return a reference to a plugin when requested, so it doesn't deal with data processing itself. We may build advanced systems that can control data flow execution later (this system is going to be flexible enough), but at the moment it doesn't seem necessary

---
In reply to: [194565945](https://github.com/PlayFab/SDKGenerator/pull/134#discussion_r194565945) [](ancestors = 194565945)

Reply:
I added a few paragraphs below in this spec that explain this in details.

---
In reply to: [194568948](https://github.com/PlayFab/SDKGenerator/pull/134#discussion_r194568948) [](ancestors = 194568948,194565945)